src/bidiMapper/modules/session/SubscriptionManager.ts

```diff
@@ -232,164 +189,189 @@ export class SubscriptionManager {
-    }
-
-    return affectedContextIds.map((contextId) => ({
-      event,
-      contextId,
-    }));
+    const subscription: Subscription = {
+      id: uuidv4(),
+      eventNames: new Set(unrollEvents(eventNames)),
+      topLevelTraversableIds: new Set(
+        contextIds.map((contextId) => {
+          const topLevelContext =
+            this.#browsingContextStorage.findTopLevelContextId(contextId);
+          if (!topLevelContext) {
+            throw new NoSuchFrameException(
+              `Top-level navigable not found for context id ${contextId}`,
+            );
+          }
+          return topLevelContext;
+        }),
+      ),
+      channel,
+    };
+    this.#subscriptions.push(subscription);
+    return subscription;
   }
 
   /**
    * Unsubscribes atomically from all events in the given contexts and channel.
+   *
+   * This is a legacy spec branch to unsubscribe by attributes.
    */
-  unsubscribeAll(
-    events: ChromiumBidi.EventNames[],
-    contextIds: (BrowsingContext.BrowsingContext | null)[],
+  unsubscribe(
+    inputEventNames: ChromiumBidi.EventNames[],
+    inputContextIds: BrowsingContext.BrowsingContext[],
     channel: BidiPlusChannel,
   ) {
-    // Assert all contexts are known.
-    for (const contextId of contextIds) {
-      if (contextId !== null) {
-        this.#browsingContextStorage.getContext(contextId);
+    const eventNames = new Set(unrollEvents(inputEventNames));
+
+    for (const contextId of inputContextIds) {
+      // Validation that contexts exist.
+      this.#browsingContextStorage.getContext(contextId);
+    }
+    const topLevelTraversables = new Set(
+      inputContextIds.map((contextId) => {
+        const topLevelContext =
+          this.#browsingContextStorage.findTopLevelContextId(contextId);
+        if (!topLevelContext) {
+          throw new NoSuchFrameException(
+            `Top-level navigable not found for context id ${contextId}`,
+          );
+        }
+        return topLevelContext;
+      }),
+    );
+
+    const newSubscriptions: Subscription[] = [];
+    let found = false;
+    for (const subscription of this.#subscriptions) {
+      if (found || subscription.channel !== channel) {
+        newSubscriptions.push(subscription);
+        continue;
+      }
+      if (intersection(subscription.eventNames, eventNames).size === 0) {
+        newSubscriptions.push(subscription);
+        continue;
+      }
+      if (topLevelTraversables.size === 0) {
+        // If it is non-global subscription, keep it as is because we
+        // are only unsubscribing globally.
+        if (subscription.topLevelTraversableIds.size !== 0) {
+          newSubscriptions.push(subscription);
+          continue;
+        }
+        const subscriptionEventNames = new Set(subscription.eventNames);
+        for (const eventName of eventNames) {
+          if (subscriptionEventNames.has(eventName)) {
+            eventNames.delete(eventName);
```

Comment:
I'm failing to understand something: if we have 2 subscriptions with the same `eventName`, will not it be removed only from the first subscription?

Reply:
Can you please add the following test:
1. Subscribe to (a, b, c).
2. Subscribe to (b, c, d).
3. Unsubscribe from (b).
4. Assert no subscriptions to `b` left.

Reply:
Yeah, that case would not work yet.

Reply:
Let's see maybe it will work now, I wonder if the same needs to be done for non-global branch.

Reply:
Yeah, I think the other branch needs the same treatment.